File: crawler/python/get_all_match_details.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created by Williamhiler on 2024-12-13
读取指定赛季目录下的文件，获取所有比赛信息
遍历获取每场比赛的详情，并保存为round-matchId名称的文件
添加延时机制以避免被识别为爬虫
"""

import asyncio
import os
import json
import random
from get_match_detail import get_match_detail

# 延时函数，随机生成延时时间，避免被识别为爬虫
async def delay(min_ms, max_ms):
    delay_time = random.randint(min_ms, max_ms)
    print(f'等待 {delay_time} 毫秒后继续...')
    await asyncio.sleep(delay_time / 1000)

# 读取指定赛季目录下的文件，获取所有比赛信息
async def read_season_data(season):
    season_dir = os.path.join(os.path.dirname(__file__), 'output', season)
    
    # 检查赛季目录是否存在
    if not os.path.exists(season_dir):
        raise FileNotFoundError(f'赛季目录 {season_dir} 不存在')
    
    # 读取赛季目录下的所有文件
    files = os.listdir(season_dir)
    
    # 过滤出JSON文件
    json_files = [file for file in files if file.endswith('.json')]
    
    if not json_files:
        raise FileNotFoundError(f'赛季目录 {season_dir} 下没有JSON文件')
    
    # 读取第一个JSON文件（假设每个赛季只有一个JSON文件）
    data_file = json_files[0]
    data_file_path = os.path.join(season_dir, data_file)
    
    print(f'正在读取赛季数据文件: {data_file_path}')
    
    with open(data_file_path, 'r', encoding='utf-8') as f:
        match_data = json.load(f)
    
    return match_data

# 为指定比赛创建详细信息目录
def create_match_detail_dir(season, round_num, match_id):
    # 创建赛季详细信息目录
    season_detail_dir = os.path.join(os.path.dirname(__file__), 'output', season, 'details')
    if not os.path.exists(season_detail_dir):
        os.makedirs(season_detail_dir)
    
    # 创建轮次目录
    round_dir = os.path.join(season_detail_dir, round_num)
    if not os.path.exists(round_dir):
        os.makedirs(round_dir)
    
    return round_dir

# 主函数，读取赛季数据并遍历获取每场比赛的详情
async def main():
    season = '2015-2016'  # 要处理的赛季
    processed_matches = 0
    failed_matches = 0
    failed_matches_list = []
    
    try:
        print(f'开始处理 {season} 赛季的所有比赛详情...')
        
        # 读取赛季数据
        match_data = await read_season_data(season)
        total_matches = len(match_data)
        
        print(f'成功读取赛季数据，共 {total_matches} 场比赛')
        
        # 遍历所有比赛，获取详细信息
        index = 0
        for match_id, match_info in match_data.items():
            index += 1
            round_num = match_info['round']
            
            print(f'\n正在处理第 {index}/{total_matches} 场比赛')
            print(f'比赛ID: {match_id}, 轮次: {round_num}')
            print(f'主队: {match_info.get("homeTeamName", "未知")}, 客队: {match_info.get("awayTeamName", "未知")}')
            
            try:
                # 创建比赛详细信息目录
                round_dir = create_match_detail_dir(season, round_num, match_id)
                
                # 生成文件名：round-matchId.json
                file_name = f'{match_id}.json'
                file_path = os.path.join(round_dir, file_name)
                
                # 检查文件是否已存在，如果存在则跳过
                if os.path.exists(file_path):
                    print(f'文件已存在，跳过抓取: {file_path}')
                    processed_matches += 1
                else:
                    # 获取比赛详情并保存到指定路径，传入赛季信息
                    detail = await get_match_detail(match_id, file_path, season)
                
                if detail:
                    processed_matches += 1
                else:
                    print('获取比赛详情失败')
                    failed_matches += 1
                    failed_matches_list.append({'matchId': match_id, 'round': round_num})

                # 如果不是最后一场比赛，添加延时
                if index < total_matches:
                    # 添加3-5秒的随机延时，避免被识别为爬虫
                    await delay(500, 2000)
            
            except Exception as error:
                print(f'处理比赛 {match_id} 时出错: {error}')
                failed_matches += 1
                failed_matches_list.append({'matchId': match_id, 'round': round_num})
                
                # 即使出错也继续处理下一场比赛，并添加延时
                # if index < total_matches:
                #     await delay(1000, 2000)  # 出错后使用较短的延时
        
        print(f'\n=== 处理完成 ===')
        print(f'总比赛数: {total_matches}')
        print(f'成功处理: {processed_matches}')
        print(f'处理失败: {failed_matches}')
        if failed_matches_list:
            print(f'\n失败的比赛记录:')
            for i, failed_match in enumerate(failed_matches_list, 1):
                print(f'{i}. 轮次: {failed_match["round"]}, 比赛ID: {failed_match["matchId"]}')
            print(f'\n失败记录总数: {len(failed_matches_list)}')
        print(f'================')
        
    except Exception as error:
        print(f'执行过程中发生错误: {error}')
        import traceback
        traceback.print_exc()

# 启动主函数
if __name__ == "__main__":
    asyncio.run(main())

File: crawler/js/getAllMatchDetails.js
/**
 * 读取指定赛季目录下的文件，获取所有比赛信息
 * 遍历获取每场比赛的详情，并保存为round-matchId名称的文件
 * 添加延时机制以避免被识别为爬虫
 */

var getMatchDetail = require("./getMatchDetail");
var fs = require('fs');
var path = require('path');

/**
 * 延时函数，随机生成延时时间，避免被识别为爬虫
 * @param min 最小延时时间（毫秒）
 * @param max 最大延时时间（毫秒）
 * @returns {Promise<void>}
 */
function delay(min, max) {
    const delayTime = Math.floor(Math.random() * (max - min + 1)) + min;
    console.log(`等待 ${delayTime} 毫秒后继续...`);
    return new Promise(resolve => setTimeout(resolve, delayTime));
}

/**
 * 读取指定赛季目录下的文件，获取所有比赛信息
 * @param season 赛季名称，如"2024-2025"
 * @returns {Promise<Object>}
 */
function readSeasonData(season) {
    return new Promise((resolve, reject) => {
        const seasonDir = path.join(__dirname, 'output', season);
        
        // 检查赛季目录是否存在
        if (!fs.existsSync(seasonDir)) {
            reject(new Error(`赛季目录 ${seasonDir} 不存在`));
            return;
        }
        
        // 读取赛季目录下的所有文件
        fs.readdir(seasonDir, (err, files) => {
            if (err) {
                reject(err);
                return;
            }
            
            // 过滤出JSON文件
            const jsonFiles = files.filter(file => file.endsWith('.json'));
            
            if (jsonFiles.length === 0) {
                reject(new Error(`赛季目录 ${seasonDir} 下没有JSON文件`));
                return;
            }
            
            // 读取第一个JSON文件（假设每个赛季只有一个JSON文件）
            const dataFile = jsonFiles[0];
            const dataFilePath = path.join(seasonDir, dataFile);
            
            console.log(`正在读取赛季数据文件: ${dataFilePath}`);
            
            fs.readFile(dataFilePath, 'utf8', (err, data) => {
                if (err) {
                    reject(err);
                    return;
                }
                
                try {
                    const matchData = JSON.parse(data);
                    resolve(matchData);
                } catch (parseError) {
                    reject(new Error(`解析JSON文件 ${dataFilePath} 失败: ${parseError.message}`));
                }
            });
        });
    });
}

/**
 * 为指定比赛创建详细信息目录
 * @param season 赛季名称
 * @param round 轮次
 * @param matchId 比赛ID
 * @returns {string} 创建的目录路径
 */
function createMatchDetailDir(season, round, matchId) {
    // 创建赛季详细信息目录
    const seasonDetailDir = path.join(__dirname, 'output', season, 'details');
    if (!fs.existsSync(seasonDetailDir)) {
        fs.mkdirSync(seasonDetailDir);
    }
    
    // 创建轮次目录
    const roundDir = path.join(seasonDetailDir, round);
    if (!fs.existsSync(roundDir)) {
        fs.mkdirSync(roundDir);
    }
    
    return roundDir;
}

/**
 * 主函数，读取赛季数据并遍历获取每场比赛的详情
 */
async function main() {
    const season = '2015-2016'; // 要处理的赛季
    let processedMatches = 0;
    let failedMatches = 0;
    let failedMatchesList = [];
    
    try {
        console.log(`开始处理 ${season} 赛季的所有比赛详情...`);
        
        // 读取赛季数据
        const matchData = await readSeasonData(season);
        const totalMatches = Object.keys(matchData).length;
        
        console.log(`成功读取赛季数据，共 ${totalMatches} 场比赛`);
        
        // 遍历所有比赛，获取详细信息
        let index = 0;
        for (const matchId in matchData) {
            if (matchData.hasOwnProperty(matchId)) {
                const matchInfo = matchData[matchId];
                const round = matchInfo.round;
                
                index++;
                console.log(`\n正在处理第 ${index}/${totalMatches} 场比赛`);
                console.log(`比赛ID: ${matchId}, 轮次: ${round}`);
                console.log(`主队: ${matchInfo.homeTeamName}, 客队: ${matchInfo.awayTeamName}`);
                
                try {
                    // 创建比赛详细信息目录
                    const roundDir = createMatchDetailDir(season, round, matchId);
                    
                    // 生成文件名：round-matchId.json
                    const fileName = `${matchId}.json`;
                    const filePath = path.join(roundDir, fileName);
                    
                    // 检查文件是否已存在，如果存在则跳过
                    if (fs.existsSync(filePath)) {
                        console.log(`文件已存在，跳过抓取: ${filePath}`);
                        processedMatches++;
                    } else {
                        // 获取比赛详情并保存到指定路径，传入赛季信息
                            const detail = await getMatchDetail(matchId, filePath, season);
                        
                        if (detail) {
                            processedMatches++;
                        } else {
                            console.log(`获取比赛详情失败`);
                            failedMatches++;
                            failedMatchesList.push({ matchId, round });
                        }

                        // 如果不是最后一场比赛，添加延时
                        if (index < totalMatches) {
                            // 添加3-5秒的随机延时，避免被识别为爬虫
                            await delay(500, 2000);
                        }
                    }
                    

                    
                } catch (error) {
                    console.error(`处理比赛 ${matchId} 时出错:`, error.message);
                    failedMatches++;
                    failedMatchesList.push({ matchId, round });
                    
                    // 即使出错也继续处理下一场比赛，并添加延时
                    // if (index < totalMatches) {
                    //     await delay(1000, 2000); // 出错后使用较短的延时
                    // }
                }
            }
        }
        
        console.log(`\n=== 处理完成 ===`);
        console.log(`总比赛数: ${totalMatches}`);
        console.log(`成功处理: ${processedMatches}`);
        console.log(`处理失败: ${failedMatches}`);
        if (failedMatchesList.length > 0) {
            console.log(`\n失败的比赛记录:`);
            failedMatchesList.forEach((failedMatch, index) => {
                console.log(`${index + 1}. 轮次: ${failedMatch.round}, 比赛ID: ${failedMatch.matchId}`);
            });
            console.log(`\n失败记录总数: ${failedMatchesList.length}`);
        }
        console.log(`================`);
        
    } catch (error) {
        console.error(`执行过程中发生错误:`, error.message);
    }
}

// 启动主函数
main().catch(error => {
    console.error('执行过程中发生错误:', error);
});
